File: Data-Structures-and-Algorithms/0209-MinimizeSizeSubarraySum/minSubArrayLen.js
const minSubArrayLen = function(target, nums) {
    let left = 0;
    let right = 0;
    let cursum = 0;
    let ans = Infinity;

    for (var i=0; i < nums.length; i++) {
        cursum += nums[i]
        right = i;
        while (cursum >= target) {
            ans = Math.min(right - left + 1, ans)
            cursum -= nums[left];
            left ++;
        }
    }
    return (ans == Infinity)? 0 : ans
};

// O(n) single iteration through n elements
// O(1) for space (constant extra space)
// Return minimal length of continuous elements with sum bigger and equal to the target

File: Data-Structures-and-Algorithms/0209-MinimizeSizeSubarraySum/minSubArrayLen.py
class Solution(object):
    def minSubArrayLen(self, nums, target):
        ans = float("inf")
        cursum = 0
        left = 0
        right = 0
        for i in range(len(nums)):
            right = i   
            cursum += nums[i]
            while cursum >= target and left <= right:
                ans = min(right - left + 1, ans)
                cursum -= nums[left]
                left += 1       
        return 0 if ans == float("inf") else ans

nums = [2, 3, 1, 2, 4, 3, 1, 2, 1, 2, 4]
target = 500
print(Solution().minSubArrayLen(nums, target))

# O(n) single iteration through n elements
# O(1) for space (constant extra space)
# Return minimal length of continuous elements with sum bigger and equal to the target
